src/EditorFeatures/TestUtilities/Threading/StaTaskScheduler.cs

```diff
@@ -55,6 +58,34 @@ public StaTaskScheduler()
                 DispatcherSynchronizationContext = synchronizationContext;
             }
 
+            // We've created an STA thread, which has some extra requirements for COM Runtime
+            // Callable Wrappers (RCWs). If any COM object is created on the STA thread, calls to that
+            // object must be made from that thread; when the RCW is no longer being used by any
+            // managed code, the RCW is put into the finalizer queue, but to actually finalize it
+            // it has to marshal to the STA thread to do the work. This means that in order to safely
+            // clean up any RCWs, we need to ensure that the thread is pumping past the point of
+            // all RCWs being finalized
+            //
+            // This constraint is particularly problematic if our tests are running in an AppDomain:
+            // when the AppDomain is unloaded, any threads (including our STA thread) are going to be
+            // aborted. Once the thread and AppDomain is being torn down, the CLR is going to try cleaning up
+            // any RCWs associated them, because if the thread is gone for good there's no way
+            // it could ever clean anything further up. The code there waits for the finalizer queue
+            // -- but the finalizer queue might be already trying to clean up an RCW, which is marshaling
+            // to the STA thread. This could then deadlock.
+            //
+            // The suggested workaround from the CLR team is to do an explicit GC.Collect and
+            // WaitForPendingFinalizers before we let the AppDomain shut down. The belief is subscribing
+            // to the Unload is a reasonable place to do it.
+            if (Interlocked.CompareExchange(ref s_SubscribedToAppDomainUnload, 1, 0) == 0)
+            {
+                AppDomain.CurrentDomain.DomainUnload += (sender, e) =>
+                {
+                    GC.Collect();
```

Comment:
Most of the time I see the `Collect` + `WaitForPendingFinalizers` pattern it's done in a loop. Assume there is a reason you didn't need to do that here. Curious about what it would be.

Reply:
I wasn't sure how much to loop. So, ¯\_(ツ)_/¯?

Reply:
The magic number I'm thinking of here is three. I have no idea why I I think that though.

Reply:
Tradition dictates three, but that's cargo cult programming if I've ever seen it.

Reply:
I can't argue with that.

Reply:
[The number of the counting shall be...3](https://www.youtube.com/watch?v=-IOMNUayJjI)